File: typescript/src/core/xbrl/rendered-reports.ts
/**
 * Reader for a filing's rendered reports.
 *
 * Every XBRL filing ships with `FilingSummary.xml` (the list of rendered
 * reports, with roles and long names) and one `R<n>.htm` file per report
 * (the rendered statement table). This module handles fetching and
 * caching; the actual HTML parsing lives in `report-html-parser`.
 */

import { parseReportHtml, StatementLineItem } from "./report-html-parser";

export interface ReportDescriptor {
  definition: string;
  shortName: string;
  role: string;
  rFile: string;
  reportType: string;
}

interface RawHttpClient {
  getRaw(url: string): Promise<string>;
}

function findXmlText(block: string, tag: string): string {
  const match = block.match(new RegExp(`<${tag}>([\\s\\S]*?)</${tag}>`));
  return match ? match[1].trim() : "";
}

/**
 * Lists and parses a filing's rendered reports.
 *
 * Only depends on an HTTP client (`getRaw(url)`) and the filing's
 * archive-folder base URL, so it is reusable outside `XBRLInstance`.
 */
export class RenderedReportReader {
  private readonly archiveBase: string;
  private readonly http: RawHttpClient;
  private reports: ReportDescriptor[] | null = null;
  private readonly statementCache = new Map<string, StatementLineItem[]>();

  constructor(archiveBase: string, http: RawHttpClient) {
    this.archiveBase = archiveBase;
    this.http = http;
  }

  /** Parse FilingSummary.xml into a list of report descriptors. */
  async listReports(): Promise<ReportDescriptor[]> {
    if (this.reports !== null) {
      return this.reports;
    }

    let raw: string;
    try {
      raw = await this.http.getRaw(`${this.archiveBase}/FilingSummary.xml`);
    } catch {
      this.reports = [];
      return this.reports;
    }

    const reports: ReportDescriptor[] = [];
    const reportPattern = /<Report[^>]*>([\s\S]*?)<\/Report>/g;
    let match;
    while ((match = reportPattern.exec(raw)) !== null) {
      const block = match[1];
      const descriptor: ReportDescriptor = {
        definition: findXmlText(block, "LongName"),
        shortName: findXmlText(block, "ShortName"),
        role: findXmlText(block, "Role"),
        rFile: findXmlText(block, "HtmlFileName"),
        reportType: findXmlText(block, "ReportType"),
      };
      if (descriptor.role || descriptor.rFile) {
        reports.push(descriptor);
      }
    }

    this.reports = reports;
    return reports;
  }

  /** Find a report by role URI, R-file name, or short name. */
  async findReport(role: string): Promise<ReportDescriptor | null> {
    const reports = await this.listReports();
    for (const report of reports) {
      if (
        role === report.role ||
        role === report.rFile ||
        role === report.shortName
      ) {
        return report;
      }
    }
    const roleLower = role.toLowerCase();
    for (const report of reports) {
      if (report.role.toLowerCase().endsWith(roleLower)) {
        return report;
      }
    }
    return null;
  }

  /**
   * Parse one rendered report into line items.
   *
   * @param role The report's role URI (from `listReports()`), its R-file
   *   name, or its short name.
   */
  async readStatement(role: string): Promise<StatementLineItem[]> {
    const report = await this.findReport(role);
    if (!report || !report.rFile) {
      return [];
    }

    const cached = this.statementCache.get(report.rFile);
    if (cached) {
      return cached;
    }

    let raw: string;
    try {
      raw = await this.http.getRaw(`${this.archiveBase}/${report.rFile}`);
    } catch {
      return [];
    }

    const items = parseReportHtml(raw);
    this.statementCache.set(report.rFile, items);
    return items;
  }
}

File: python/src/sec_edgar_toolkit/core/xbrl/rendered_reports.py
"""
Reader for a filing's rendered reports.

Every XBRL filing ships with ``FilingSummary.xml`` (the list of rendered
reports, with roles and long names) and one ``R<n>.htm`` file per report
(the rendered statement table). This module handles fetching and
caching; the actual HTML parsing lives in :mod:`.report_html` and needs
no third-party dependencies.
"""

from __future__ import annotations

import logging
from typing import Any, Dict, List, Optional

from .report_html import (  # noqa: F401 - re-exported for backward compatibility
    normalize_period_label,
    parse_report_html,
    parse_report_number,
)

logger = logging.getLogger(__name__)


class RenderedReportReader:
    """
    Lists and parses a filing's rendered reports.

    Only depends on an HTTP client (``get_raw(url) -> bytes``) and the
    filing's archive-folder base URL, so it is reusable outside
    ``XBRLInstance``.
    """

    def __init__(self, archive_base: str, http_client: Any) -> None:
        self._archive_base = archive_base
        self._http = http_client
        self._reports: Optional[List[Dict[str, Any]]] = None
        self._statement_cache: Dict[str, List[Dict[str, Any]]] = {}

    def list_reports(self) -> List[Dict[str, Any]]:
        """Parse FilingSummary.xml into a list of report descriptors."""
        if self._reports is not None:
            return self._reports

        import xml.etree.ElementTree as ET

        url = f"{self._archive_base}/FilingSummary.xml"
        try:
            raw = self._http.get_raw(url)
        except Exception as e:
            logger.warning(f"No FilingSummary.xml at {url}: {e}")
            self._reports = []
            return self._reports

        reports: List[Dict[str, Any]] = []
        try:
            root = ET.fromstring(raw)
            for report in root.iter("Report"):
                descriptor = {
                    "definition": (report.findtext("LongName") or "").strip(),
                    "short_name": (report.findtext("ShortName") or "").strip(),
                    "role": (report.findtext("Role") or "").strip(),
                    "r_file": (report.findtext("HtmlFileName") or "").strip(),
                    "report_type": (report.findtext("ReportType") or "").strip(),
                }
                if descriptor["role"] or descriptor["r_file"]:
                    reports.append(descriptor)
        except ET.ParseError as e:
            logger.warning(f"Could not parse FilingSummary.xml: {e}")

        self._reports = reports
        return reports

    def find_report(self, role: str) -> Optional[Dict[str, Any]]:
        """Find a report by role URI, R-file name, or short name."""
        for report in self.list_reports():
            if role in (report["role"], report["r_file"], report["short_name"]):
                return report
        # Loose match on the role suffix
        role_lower = role.lower()
        for report in self.list_reports():
            if report["role"].lower().endswith(role_lower):
                return report
        return None

    def read_statement(self, role: str) -> List[Dict[str, Any]]:
        """
        Parse one rendered report into line items.

        Args:
            role: The report's role URI (from ``list_reports()``), its
                R-file name, or its short name.

        Returns:
            List of line-item dicts with ``concept``, ``label``,
            ``section``, ``values`` (period -> number), ``units``
            (period -> unit), and ``has_values`` keys.
        """
        report = self.find_report(role)
        if not report or not report.get("r_file"):
            return []

        cache_key = report["r_file"]
        if cache_key in self._statement_cache:
            return self._statement_cache[cache_key]

        url = f"{self._archive_base}/{report['r_file']}"
        try:
            raw = self._http.get_raw(url)
        except Exception as e:
            logger.warning(f"Could not fetch report {url}: {e}")
            return []

        items = self._parse_report_html(raw)
        self._statement_cache[cache_key] = items
        return items

    def _parse_report_html(self, raw: bytes) -> List[Dict[str, Any]]:
        """Parse an R<n>.htm rendered report table into line items."""
        return parse_report_html(raw)
